### PR TITLE
Resolves #56, Resolves #59, chutzpah and folder parsers added

### LINKED ISSUE #56: Support for Chutzpah coverage (java script)
I'm using visual studio to write my c# code, I also write my web application code with typescript and Chutzpah.
Is it possible to add support for Chutzpah coverage report as well? I think it can write a few output format for coverage as far as I can understand from their wiki:
https://github.com/mmanela/chutzpah/wiki/command-line-options

### LINKED ISSUE #59: Support multiple files
Consider the following scenario: a user wants to send multiple coverage files to coveralls website from a single command line run.
Feature proposal: add `--folder` flag.
This flag will receive a folder as input.
The code will take all the files in that folder and parse them according to file extensions.
file.opencover.xml (or file.opencover haven't decided yet) will be parsed as open cover
file.chuzpah.json (or file.chutzpah) will be parsed as chutzpah
file.lcov will be parsed as lcov (pending implementation of lcov format).

I'm not entirely familiar with the other options but I think that there's an option now that uses a folder which may use the same mechanism.

As a part of this changes I really think it would be wise to add an IoC container and some folders to split the logic into business logic and data access.
The current state doesn't really support unit testing well as far as I could see.

Also I think that the parsers should receive a file path and they should all inherit from the same interface.

What do you think?

### PR DESCRIPTION
**Pull request by HarelM**
Refactoring of classes and added the ability to parse an entire folder and a chutzpah.json file.
I think this pull request targets the develop branch as you asked.
Most of the code remained as is in terms of logic, I moved a few files around and added some basic functionality to reduce code duplication.

### CODE REVIEW

File: src/csmacnz.Coveralls.Tests/TestFileSystem.cs [hunk 1]
```diff
@@ -1,24 +0,0 @@
-using System.Collections.Generic;
-using Beefeater;
-
-namespace csmacnz.Coveralls.Tests
-{
-    public class TestFileSystem : IFileSystem
```
Comment (csMACnz) [inline src/csmacnz.Coveralls.Tests/TestFileSystem.cs]: I'm not happy you deleted this in place for NSubstitute. Can you roll it back please? and remove the NSubstitute dependency?

**Comment by csMACnz** [comment]
I had a quick scan over. you've refactored too much for me to accept this without further reading. I can't review this in the short term, it is too many changes.

**Comment by HarelM** [comment]
What do you suggest then? 
The ```TestFileSystem``` class was not in use before I deleted it. The unit tests I made require too much logic to be inserted to this class to properly pass the tests so I needed a mocking framework. If you want me to use a different one I don't mind, but keeping this class is just a recipe for bugs in the tests...

**Comment by HarelM** [comment]
P.S. the changes are not so big it just looks scary, I can walk you through it if you'd like.

**Comment by HarelM** [comment]
Maybe we can merge it to a different branch, something like release-0.8 and you could create a pre-release NuGet package so I'll be able to test it on my CI server to see that it is working as expected?

**Comment by csMACnz** [comment]
@HarelM I've take the essence of your Chutzpah support and added it [here](https://github.com/csMACnz/coveralls.net/commit/d7ab94e8d71e8c1b71cc5ed733694f28ec5ba280).
I want a more configuration over convention for the multiple files so will look at adding that next.

There was a lot about the rest of the refactoring I didn't like so couldn't merge this as is. As I said earlier, I don't necessarily disagree with the sentiment behind it. It just doesn't move the code in the direction that I plan to take it, and would create more work for me later on to include it now.

**Comment by HarelM** [comment]
Thanks for the update, As I said, the multiple files was a short "hacky" development in order not to break anything (note that after the refactoring the changes I needed to do were very small in order to add it), I didn't like that it relayed on file extensions but I can't think of a better idea that would still keep this tool user friendly as it is now and not break the current usage - more configuration is usually less user friendly, keep it in mind :-).

As for the chutzpah commit - the essence you took looks partial as there is an issue there with the names of the files: see the issue I opened here for the chutzpah owner:
https://github.com/mmanela/chutzpah/issues/546
The fix I added in the code which is now missing as far as I can tell was to find the file in the file system and change its name according to casing.
When the casing is incorrect posing to coveralls website creates an issue where the file is not found in the repository on Github in order to see the covered lines in the file itself.

I'm not sure where you want to go in terms of refactoring but the current code has a lot of duplication and IMO has too much logic in the "main" function and there aren't any architectural layers so some classes are impossible to test as they use the file system. If not anything, I would advise to place the parsers in their own folder as there will probably be more...

Keep me posted when you have a solution for multiple files, I'll be glad to test it.
Good luck!

**Comment by csMACnz** [comment]
I've got a multiple files support added, without thorough testing it seems ok. I'll try and publish to pre-releases this weekend.

The filenames should be fine. As I suspected, that piece of code shouldn't be necessary, I followed the same approach as my other parsers and it all seems to work. cases are maintained, correct line endings come through. You do need to use the relative base path options. (API V2.0 will probably make that mandatory to be relative, maybe defaulting to the execution directory, because of the GitHub path issues I think you were seeing)  Everything looks like it should work the way I have it.

I'll post back once a package is available.
Thats for the feature request anyways. Chutzpah looks like something I should have a play with anyways.

**Comment by csMACnz** [comment]
keep in mind this has has minimal testing: https://www.nuget.org/packages/coveralls.net/0.8.0-unstable0008

**Comment by HarelM** [comment]
Hi, Seems to be working as expected, thanks a lot!
Results can be found here:
https://coveralls.io/github/IsraelHikingMap/Site
powershell script can be found here:
https://github.com/IsraelHikingMap/Site/blob/master/AppveyorCIScripts/PostTests.ps1
Relevant command lines in case you are interested:
chutzpah:
```C:\projects\site\packages\Chutzpah.4.2.3\tools\chutzpah.console.exe C:\projects\site\chutzpah.json /junit chutzpah-junit.xml /coverage /coveragejson coverage-chutzpah.json```
open cover:
```C:\projects\site\packages\OpenCover.4.6.519\tools\OpenCover.Console.exe -register:user -target:"C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\IDE\CommonExtensions\Microsoft\TestWindow\vstest.console.exe" -targetargs:"C:\projects\site\Tests\IsraelHiking.API.Tests\bin\Debug\IsraelHiking.API.Tests.dll C:\projects\site\Tests\IsraelHiking.DataAccess.Tests\bin\Debug\IsraelHiking.DataAccess.Tests.dll" -filter:"+[*]*API* +[*]*Database* +[*]*GPSBabel* -[*]*JsonResponse* -[*]*GpxTypes* -[*]*Tests*" -excludebyattribute:"*.ExcludeFromCodeCoverage*" -output:coverage-opencover.xml```
coveralls:
```C:\projects\site\packages\coveralls.net.0.8.0-unstable0008\tools\csmacnz.Coveralls.exe --multiple -i "opencover=coverage-opencover.xml;chutzpah=coverage-chutzpah.json" --repoToken w3WvP9CEZ5M23oBONNsalxIgEzOmBwo9f --commitId 554f1eeed73b68ba4d7eaaa64924ebbb58b6ac85 --commitBranch master --commitAuthor "Harel Mazor" --commitMessage "#110 - Added appveyor powershell script to add coverage to the build process." --jobId f8rkqh6j7rig4asm --commitEmail none --useRelativePaths```

**Comment by csMACnz** [comment]
working as expected through https://github.com/csMACnz/coveralls.net/commit/d7ab94e8d71e8c1b71cc5ed733694f28ec5ba280

**Pull request closed without merging**